Null bytes in regex patterns are not escaped or rejected

## Description

Fuzzing discovered that null bytes (`\x00`) in regex patterns (used by `matches()`) are not properly handled. While we fixed null bytes in string literals (#14) and LIKE patterns (#15), the regex pattern handling in `matches()` has the same vulnerability.

## Discovered By

- **CI Run**: https://github.com/SPANDigital/cel2sql/actions/runs/18402929535
- **Job ID**: 52436172023
- **Fuzzer**: `FuzzConvertRE2ToPOSIX`
- **Crash input**: `testdata/fuzz/FuzzConvertRE2ToPOSIX/390ca22614d4ce1a`

## Error

```
--- FAIL: FuzzConvertRE2ToPOSIX (0.15s)
    fuzz_test.go:345: Generated SQL contains null bytes: "email ~ '\x00'"
```

## Reproduction

```bash
go test -run=FuzzConvertRE2ToPOSIX/390ca22614d4ce1a
```

**Input pattern**: `"\x00"` (null byte)  
**CEL expression**: `email.matches(r"\x00")`  
**Generated SQL**: `email ~ '\x00'` (contains unescaped null byte)

## Impact

- **Security**: Null bytes in regex patterns can corrupt SQL queries
- **Correctness**: Pattern matching may behave unexpectedly  
- **Consistency**: String literals (#14) and LIKE patterns (#15) reject null bytes, but regex patterns don't

## Root Cause

The `callMatches()` function processes regex patterns and converts them from RE2 to POSIX format, but it doesn't check for null bytes before generating the `~` or `~*` operator SQL.

Previous fixes:
- PR #14: Fixed string literals in `visitConst()`
- PR #16: Fixed LIKE patterns in `callStartsWith()` and `callEndsWith()`
- Missing: Regex patterns in `callMatches()`

## Suggested Fix

Add null byte detection in the regex handling code before generating the pattern:

```go
func (con *converter) callMatches(...) error {
    // ... existing code to extract pattern ...
    
    pattern := args[0].GetConstExpr().GetStringValue()
    
    // Reject patterns containing null bytes
    if strings.Contains(pattern, "\x00") {
        return errors.New("regex patterns cannot contain null bytes")
    }
    
    // ... continue with RE2 to POSIX conversion ...
}
```

## Test Coverage Needed

Once fixed, add explicit regression tests for:
- `email.matches(r"\x00")` → error
- `email.matches(r"\x00test")` → error  
- `email.matches(r"test\x00")` → error
- `email.matches(r"valid.*pattern")` → success

## Related Issues

- #12 - Original null byte bug in string literals (fixed in #14)
- #15 - Null bytes in LIKE patterns (fixed in #16)
- PR #13 - Fuzzing infrastructure that discovered this issue
- PR #14 - Fixed null bytes in string literals
- PR #16 - Fixed null bytes in LIKE patterns

## Pattern

This is the third null byte vulnerability found by fuzzing, suggesting a pattern:
1. **String literals** (`name == "\x00"`) - Fixed in #14
2. **LIKE patterns** (`name.startsWith("\x00")`) - Fixed in #16
3. **Regex patterns** (`email.matches(r"\x00")`) - This issue

We should audit all other string processing paths for similar issues.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

Co-Authored-By: Claude <noreply@anthropic.com>

--- a/cel2sql.go
+++ b/cel2sql.go
@@ -409,8 +409,12 @@ func (con *converter) callMatches(target *exprpb.Expr, args []*exprpb.Expr) erro
 	if constExpr := patternExpr.GetConstExpr(); constExpr != nil && constExpr.GetStringValue() != "" {
 		// Convert RE2 pattern to POSIX
 		re2Pattern := constExpr.GetStringValue()
+		// Reject patterns containing null bytes
+		if strings.Contains(re2Pattern, "\x00") {
+			return errors.New("regex patterns cannot contain null bytes")
+		}
 		posixPattern := convertRE2ToPOSIX(re2Pattern)
-		
+
 		// Write the converted pattern as a string literal
 		escaped := strings.ReplaceAll(posixPattern, "'", "''")
 		con.str.WriteString("'")

--- a/cel2sql_test.go
+++ b/cel2sql_test.go
@@ -634,3 +634,75 @@ func TestNullByteInLikePatterns(t *testing.T) {
 		})
 	}
 }
+
+// TestNullByteInRegexPatterns tests that regex patterns with null bytes are rejected
+func TestNullByteInRegexPatterns(t *testing.T) {
+	env, err := cel.NewEnv(
+		cel.Variable("email", cel.StringType),
+	)
+	require.NoError(t, err)
+
+	tests := []struct {
+		name    string
+		expr    string
+		wantErr string
+	}{
+		{
+			name:    "matches with only null byte",
+			expr:    `email.matches("\x00")`,
+			wantErr: "regex patterns cannot contain null bytes",
+		},
+		{
+			name:    "matches with null byte at start",
+			expr:    `email.matches("\x00test")`,
+			wantErr: "regex patterns cannot contain null bytes",
+		},
+		{
+			name:    "matches with null byte at end",
+			expr:    `email.matches("test\x00")`,
+			wantErr: "regex patterns cannot contain null bytes",
+		},
+		{
+			name:    "matches with null byte in middle",
+			expr:    `email.matches("te\x00st")`,
+			wantErr: "regex patterns cannot contain null bytes",
+		},
+		{
+			name:    "matches function style with null byte",
+			expr:    `matches(email, "\x00")`,
+			wantErr: "regex patterns cannot contain null bytes",
+		},
+		{
+			name:    "matches with multiple null bytes",
+			expr:    `email.matches("\x00\x00\x00")`,
+			wantErr: "regex patterns cannot contain null bytes",
+		},
+		{
+			name:    "matches with valid pattern",
+			expr:    `email.matches(r"^[a-z]+@.*\.com$")`,
+			wantErr: "",
+		},
+		{
+			name:    "matches with valid complex pattern",
+			expr:    `email.matches(r".*")`,
+			wantErr: "",
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			ast, issues := env.Compile(tt.expr)
+			require.NoError(t, issues.Err())
+
+			got, err := cel2sql.Convert(ast)
+			if tt.wantErr != "" {
+				require.Error(t, err)
+				assert.Contains(t, err.Error(), tt.wantErr)
+			} else {
+				require.NoError(t, err)
+				// Just verify it doesn't error - actual SQL format tested elsewhere
+				assert.NotEmpty(t, got)
+			}
+		})
+	}
+}

